driver/docker/docker.go

```diff
@@ -218,25 +221,71 @@ func (d *Driver) exec(op *driver.Operation) error {
 		}
 	}()
 
-	statusc, errc := cli.Client().ContainerWait(ctx, resp.ID, container.WaitConditionRemoved)
+	statusc, errc := cli.Client().ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
 	if err = cli.Client().ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
-		return fmt.Errorf("cannot start container: %v", err)
+		return driver.OperationResult{}, fmt.Errorf("cannot start container: %v", err)
 	}
 	select {
 	case err := <-errc:
 		if err != nil {
-			return fmt.Errorf("error in container: %v", err)
+			opResult, _ := d.fetchOutputs(ctx, resp.ID)
+			return opResult, fmt.Errorf("error in container: %v", err)
 		}
 	case s := <-statusc:
 		if s.StatusCode == 0 {
-			return nil
+			return d.fetchOutputs(ctx, resp.ID)
 		}
 		if s.Error != nil {
-			return fmt.Errorf("container exit code: %d, message: %v", s.StatusCode, s.Error.Message)
+			opResult, _ := d.fetchOutputs(ctx, resp.ID)
+			return opResult, fmt.Errorf("container exit code: %d, message: %v", s.StatusCode, s.Error.Message)
 		}
-		return fmt.Errorf("container exit code: %d", s.StatusCode)
+		opResult, _ := d.fetchOutputs(ctx, resp.ID)
+		return opResult, fmt.Errorf("container exit code: %d", s.StatusCode)
+	}
+	opResult, _ := d.fetchOutputs(ctx, resp.ID)
+	return opResult, err
+}
+
+// fetchOutputs takes a context and a container ID; it copies the /cnab/app/outputs directory from that container.
+// The goal is to collect all the files in the directory (recursively) and put them in a flat map of path to contents.
+// This map will be inside the OperationResult. When fetchOutputs returns an error, it may also return partial results.
+func (d *Driver) fetchOutputs(ctx context.Context, container string) (driver.OperationResult, error) {
+	opResult := driver.OperationResult{
+		Outputs: map[string]string{},
+	}
+	ioReader, _, err := d.dockerCli.Client().CopyFromContainer(ctx, container, "/cnab/app/outputs")
+	if err != nil {
+		return opResult, fmt.Errorf("error copying outputs from container: %s", err)
 	}
-	return err
+
+	tarReader := tar.NewReader(ioReader)
+	header, err := tarReader.Next()
+
+	// io.EOF pops us out of loop on successful run.
+	for err == nil {
+		// skip directories because we're gathering file contents
+		if header.FileInfo().IsDir() {
+			header, err = tarReader.Next()
+			continue
+		}
+
+		var contents []byte
+		// CopyFromContainer strips prefix above outputs directory.
+		pathInContainer := filepath.Join("/cnab/app", header.Name)
```

Comment:
I'm wondering about the behaviour in Windows. I think we'd end up with paths like "/cnab/app\x\y\z" which seems odd. Would it be cleaner to use paths as they were in the container? If so, just use `path.Join`. (Also, note that such paths are valid Windows paths FWIW.)

Reply:
We gave this a try in the playground and couldn't find a difference between the two https://play.golang.org/p/tQVQNw1YJZR. Is there something we're missing?

Reply:
The playground is running on *IX and so `filepath.Join` and `path.Join` give the same results. On Windows, the results would be different. (BTW if `header.Name` can have embedded back-slashes, these should be converted to forward slashes.)

If `header.Name` will have embedded back-slashes on Windows, another way to make `pathInContainer` consistent would be to use:
```
pathInContainer := filepath.Join("cnab", "app", header.Name)
```

Reply:
It would be helpful if someone with a Windows system could try this PR and see how it handles outputs in multiple directories.

Reply:
Presumably some of our friends at Microsoft could help us out with that :)
@vdice @jeremyrickard @radu-matei

Reply:
I agree with @glyn about `pathInContainer := filepath.Join("cnab", "app", header.Name)` 👍

Reply:
Based on @jeremyrickard's experiments on Windows and if the variable name `pathInContainer` (where we mean a Linux container) is to be taken literally, then the path separator should be `/` and so it's probably best to use `path.Join` instead of `filepath.Join` along these lines:
```suggestion
		pathInContainer := path.Join("cnab", "app", header.Name)
```